ARC-V RHX-100 upstream patch series

MichielDerhaeg opened the pull request.

MichielDerhaeg commented on gcc/config/riscv/riscv.cc:
```diff
@@ -10573,30 +10574,6 @@ riscv_store_data_bypass_p (rtx_insn *out_insn, rtx_insn *in_insn)
   return store_data_bypass_p (out_insn, in_insn);
 }
 
-/* Implement one boolean function for each of the values of the
```
Should the declarations in the header be moved to arcv.h?

luismgsilva replied:
I know I originally placed the declarations in `arcv.h`, but looking at the existing code structure, it may be a better fit to keep them in `riscv-protos.h`

EDIT: See https://github.com/foss-for-synopsys-dwc-arc-processors/gcc/pull/192/commits/d479345f57d82fd23d84f5778ad6b94ad23e439b

MichielDerhaeg commented on gcc/config/riscv/riscv-protos.h:
```diff
@@ -860,6 +860,9 @@ extern rtx riscv_prefetch_cookie (rtx, rtx);
 
 struct riscv_tune_param;
+
+extern int riscv_get_tune_param_issue_rate (void);
```
This causes a circular dependency between riscv.cc and arcv.cc. Can we instead just pass in the issue_rate itself as a parameter where this is called?

luismgsilva replied:
Indeed. Yes.

EDIT: See https://github.com/foss-for-synopsys-dwc-arc-processors/gcc/pull/192/commits/73219a427a993c2657bd3189054da79516dbd4b4

MichielDerhaeg commented on gcc/config/riscv/riscv.cc:
```diff
@@ -4392,7 +4392,8 @@ riscv_rtx_costs (rtx x, machine_mode mode, int outer_code, int opno ATTRIBUTE_UN
     case SIGN_EXTRACT:
-      if (TARGET_XTHEADBB && outer_code == SET
+      if ((TARGET_ARCV_RHX100 || TARGET_XTHEADBB)
+	  && outer_code == SET
```
FYI, this was added for the bit-extract fusion.

MichielDerhaeg commented on gcc/config/riscv/riscv.md:
```diff
@@ -4536,6 +4595,68 @@
+     return "slli\t%0,%1,%2\n\tsrli\t%0,%0,%3";
+  }
+  [(set_attr "type" "alu_fused")]
+)
```
As far as I can tell, this fusion was never implemented as a define_insn_and_split. Might not be trivial to force these exact instructions after a split.

MichielDerhaeg commented on gcc/config/riscv/riscv.md:
```diff
@@ -4536,6 +4595,68 @@
+     return "slli\t%0,%1,%2\n\tsrli\t%0,%0,%3";
+  }
+  [(set_attr "type" "alu_fused")]
+)
```
All of these patterns used to be a defined_insn_and_split. In case we can't get rid of them, let's try to go back to define_insn_and_split (check the downstream commit history) and see what the performance looks like.

luismgsilva replied:
If we revert `madd_split_fused1` back to a `define_insn_and_split` we get 0.898% improvement.

MichielDerhaeg replied:
I ran embench before and after the revert to `define_insn_and_split`.
There were some regressions and improvements but the numbers were mostly stable except for the matmul benchmark.
Biggest regression was edn with a 2% drop.
Largest improvement was matmult with a 12% improvements. So I think we can keep this change.

MichielDerhaeg replied:
`madd_split_fused_extended` also still needs to be converted to a `define_insn_and_split`.

luismgsilva replied:
> `madd_split_fused_extended` also still needs to be converted to a `define_insn_and_split`.

Reverted in: https://github.com/foss-for-synopsys-dwc-arc-processors/gcc/pull/192/commits/191282c0e6a85b91411b0863cb280134deeb7a9f

MichielDerhaeg reviewed (commented):
I tried to split up the commits in something sensible. Didn't check whether they can be built individually though.

MichielDerhaeg commented on gcc/config/riscv/riscv.md:
```diff
@@ -4633,17 +4633,23 @@
-  [(set_attr "type" "imul_fused")]
+    DONE;
+   }"
+  [(set_attr "type" "imul")]
```
We might want to keep imul_fused.

MichielDerhaeg commented on gcc/config/riscv/riscv.md:
```diff
@@ -4633,17 +4633,23 @@
        {
-	 return "mulw\t%4,%1,%2\n\taddw\t%4,%3,%4\n\tmv\t%0,%4";
+        emit_insn (gen_mulsi3_extended (operands[4], operands[1], operands[2]));
+        emit_insn (gen_addsi3_extended (operands[0], operands[3], operands[4]));
```
Both target operand registers need to be the same or the fusion won't happen.
We do need to emit the move here.

luismgsilva commented on gcc/config/riscv/riscv.md:
```diff
@@ -4536,6 +4595,81 @@
+       }
+    DONE;
+   }"
+  [(set_attr "type" "imul")]
```
Should this also be as follows?
```suggestion
  [(set_attr "type" "imul_fused")]
```

MichielDerhaeg commented on gcc/config/riscv/riscv.md:
```diff
@@ -4610,13 +4610,13 @@
-        emit_insn (gen_mulsi3 (operands[4], operands[1], operands[2]));
-        emit_insn (gen_addsi3 (operands[0], operands[3], operands[4]));
+	emit_insn (gen_mulsi3 (operands[4], operands[1], operands[2]));
+	emit_insn (gen_addsi3 (operands[0], operands[3], operands[4]));
```
Ah, this is also wrong.

luismgsilva replied:
Thanks. Fixed.

luismgsilva replied:
Interesting with this fix https://github.com/foss-for-synopsys-dwc-arc-processors/gcc/commit/b4ce3f98a95dfb5dc26e5c5bd0aa7449ca7e0553 I get a regression of 0.5%

MichielDerhaeg commented on gcc/config/riscv/riscv-protos.h:
```diff
@@ -875,6 +875,7 @@ extern rtx riscv_prefetch_cookie (rtx, rtx);
 extern bool riscv_prefetch_offset_address_p (rtx, machine_mode);
 
 struct riscv_tune_param;
+
```
Remove

MichielDerhaeg commented on gcc/config/riscv/arcv.cc:
```diff
@@ -522,6 +548,221 @@ arcv_macro_fusion_pair_p (rtx_insn *prev, rtx_insn *curr)
+{
+  if (!TARGET_ARCV_RHX100)
+    return priority;
+
```
We already check whether the fusion was enabled in the caller.

luismgsilva replied:
Do we want to check `RISCV_FUSE_ARCV` or specifically RHX100 ? I know currently it's the same think, but for future purposes.

MichielDerhaeg replied:
I think we want most things to be gated by `RISCV_FUSE_ARCV` except for the things added by the scheduling patch (that adds reorder2, etc).
The scheduling patch adds things that are really specific about the pipeline definition but everything else is just about the fusion schedule grouping.

MichielDerhaeg commented on gcc/config/riscv/riscv-protos.h:
```diff
@@ -839,9 +839,18 @@ extern const char *th_output_move (rtx, rtx);
+extern void arcv_sched_init (void);
+extern int arcv_sched_reorder2 (rtx_insn **, int *);
+extern int arcv_sched_adjust_priority (rtx_insn *, int);
+extern int arcv_sched_adjust_cost (rtx_insn *, int, int);
```
These need to be added in the scheduling pass.

luismgsilva replied:
I might have messed up during rebasing. Fixing it...

luismgsilva replied:
Fixed. Thanks.

MichielDerhaeg committed:
Add GitHub CI workflows for selected targets

Signed-off-by: Michiel Derhaeg <michiel.derhaeg@gf.com>

luismgsilva commented:
@MichielDerhaeg Take a look at the current fixups

MichielDerhaeg commented on gcc/config/riscv/riscv-fusion.cc:
```diff
@@ -758,5 +1001,209 @@ riscv_macro_fusion_pair_p (rtx_insn *prev, rtx_insn *curr)
+	   fprintf (dump_file, "RISCV_FUSE_SHIFT_BITEXTRACT\n");
+	 return true;
+       }
+    }
```
As far as I can tell, this is just identical to `RISCV_FUSE_BFEXT`

luismgsilva replied:
Youre correct. Please see: https://github.com/foss-for-synopsys-dwc-arc-processors/gcc/pull/192/changes/6084a0edbae7e1b8bad69a1b92db25fb5c4c71c4

MichielDerhaeg commented on gcc/config/riscv/riscv-fusion.cc:
```diff
@@ -126,6 +162,213 @@ riscv_set_is_shNadduw (rtx set)
+
+  return false;
+}
+
```
Upstream covers ld+lui already.
We can split this up so we just have to add st+lui.

luismgsilva replied:
Ive simplified and removed the function call. I believe its easier to read like this: https://github.com/foss-for-synopsys-dwc-arc-processors/gcc/pull/192/changes/5d4f626891fed8f49e39f1d645474ca1f238d178

MichielDerhaeg commented on gcc/config/riscv/riscv-fusion.cc:
```diff
@@ -758,5 +1001,209 @@ riscv_macro_fusion_pair_p (rtx_insn *prev, rtx_insn *curr)
+	 return true;
+       }
+    }
+
```
If we merge part of this fusion with the upstream version. We can outline the upstream version and add a `_REV` fusion type for the reversed case.

luismgsilva replied:
I have created a new `RISCV_FUSE_LUI_LD_REV` alongside the other `RISCV_FUSE_LUI_LD` See https://github.com/foss-for-synopsys-dwc-arc-processors/gcc/pull/192/changes/5ffb20e41305247cd9c35a8656d693c421f47b4e

But, shouldnt we align with the others? Either adapt this to use `(curr, prev)` or make it more modular and create new fusion types for the reversed cases. e.g., RISCV_FUSE_STORE_LUI and RISCV_FUSE_STORE_LUI_REV

MichielDerhaeg replied:
I would outline the original fusions, and just reverse the input params for _rev.

MichielDerhaeg commented on gcc/config/riscv/riscv-fusion.cc:
```diff
@@ -758,5 +1001,209 @@ riscv_macro_fusion_pair_p (rtx_insn *prev, rtx_insn *curr)
+	   fprintf (dump_file, "RISCV_FUSE_LS_UPDATE (curr, prev)\n");
+	 return true;
+       }
+    }
```
There are upstream ones that looks similar. Have you checked what the exact differences are? (except for the reversibility)

luismgsilva replied:
Which ones are you referring to ?

MichielDerhaeg replied:
e.g. `RISCV_FUSE_ALIGNED_STD`

luismgsilva replied:
ALIGNED_STD is stores-only. We could only add the load-only version.

But alignment requirement is different. ALIGNED_STD requires a lower offset to be aligned to `2 * mode_size`, whereas we only need each individual access to be aligned (`offset % mode_size`), so ALIGNED_STD would reject valid pairs. e.g., two sw at offsets 4 and 8 off the same base.

RISCV_FUSE_ALIGNED_STD:
```
	      /* We've normalized, so we need to check that the lower
		 address is aligned to 2X the size of the object.  The
		 higher address must be the lower address plus the
		 size of the object.  */
	      if (((offset_prev_int % (2 * mode_size)) == 0)
		  && offset_prev_int + mode_size == offset_curr_int)
```

RISCV_FUSE_ADJACENT_STORE:
```
  /* Fuse adjacent aligned addresses.  */
  if ((off0 % GET_MODE_SIZE (GET_MODE (mem0)).to_constant () == 0)
      && (abs (off1 - off0) == GET_MODE_SIZE (GET_MODE (mem0)).to_constant ()))
    return true;
```

Another thing is that ALIGNED_STD is gated on SCALAR_INT_MODE_P, which for now is fine. But we do have non-scalar modes in the future as per your PR https://github.com/foss-for-synopsys-dwc-arc-processors/gcc/pull/248.

```
	  && SCALAR_INT_MODE_P (GET_MODE (SET_DEST (curr_set)))
```

MichielDerhaeg replied:
Okay, let's keep ours.

MichielDerhaeg commented:
@exur00 not sure if this happened already, but the next_fusible_insn has to be fixed here as well

exur00 commented:
> @exur00 not sure if this happened already, but the next_fusible_insn has to be fixed here as well

I have added it

MichielDerhaeg commented on gcc/config/riscv/riscv-protos.h:
```diff
@@ -837,7 +837,7 @@ enum riscv_fusion_pairs
   RISCV_FUSE_ADJACENT_STORE = (1 << 17),
   RISCV_FUSE_LS_UPDATE = (1 << 18),
-  RISCV_FUSE_MEMOP_LUI = (1 << 19),
+  RISCV_FUSE_STORE_LUI = (1 << 19),
```
I would call it `RISCV_FUSE_LUI_ST` to align with the old ones.

MichielDerhaeg commented on gcc/testsuite/gcc.target/riscv/arcv-fusion-store-lui-dump.c:
```diff
@@ -0,0 +1,20 @@
+}
+
+/* { dg-final { scan-rtl-dump "RISCV_FUSE_STORE_LUI \\(prev, curr\\)" "sched2" } } */
+/* { dg-final { scan-rtl-dump "RISCV_FUSE_STORE_LUI \\(curr, prev\\)" "sched2" } } */
```
Why the changes here?

luismgsilva replied:
`RISCV_FUSE_MEMOP_LUI` used to handle both load and store LUI fusion. When we split them apart, the existing test case didnt cover `RISCV_FUSE_LUI_ST` (previously `RISCV_FUSE_STORE_LUI`), so I added a test case for it.

MichielDerhaeg commented on gcc/config/riscv/riscv-fusion.cc:
```diff
@@ -0,0 +1,1204 @@
+	      fprintf (dump_file, "RISCV_FUSE_LI_STORE (subreg)\n");
+	    else
+	      fprintf (dump_file, "RISCV_FUSE_LI_STORE\n");
+	  }
```
What's the point of this?

luismgsilva replied:
This was already in the original code. Just distinguishes in the dump whether the `li` and `store` operate on the same mode or not. (e..g, `li` sets a `DImode` reg but the store uses it as `SImode`). Happy to drop the distinction and just print `RISCV_FUSE_LI_STORE` if you think it's just noise.

MichielDerhaeg replied:
It can go, I think it isn't useful.

MichielDerhaeg commented on gcc/config/riscv/riscv-fusion.cc:
```diff
@@ -0,0 +1,1204 @@
+
+      if (REG_P (store_src) && REG_P (load_dest)
+	  && REGNO (store_src) == REGNO (load_dest))
+       {
```
I would merge all of the conditions in a single if statement like the other ones.

it is fine to repeat `SET_SRC (curr_set)` and such.

exur00 committed:
Add lint check CI GH actions

Performs check_GNU_style and dg-lint on the changes in the PR.

MichielDerhaeg committed:
Update CI infra with new targets and linter fix

luismgsilva commented on gcc/config/riscv/riscv-fusion.cc:
```diff
@@ -741,6 +1222,22 @@ static const struct riscv_fusion_entry riscv_fusion_table[] =
+    riscv_fuse_ls_update, "RISCV_FUSE_LS_UPDATE" },
+  { RISCV_FUSE_LUI_ST,
+    riscv_fuse_lui_st, "RISCV_FUSE_LUI_ST" },
+  { RISCV_FUSE_LI_STORE,
```
Should we keep `_STORE` or `_ST` ?

luismgsilva commented on gcc/config/riscv/arcv.cc:
```diff
@@ -0,0 +1,417 @@
+       && (insn_type == TYPE_LOAD || insn_type == TYPE_STORE))
+    {
+      sched_state.pipeB_scheduled_p = 1;
+      sched_state.cached_can_issue_more = 1;
```
For some reason `sched_state.cached_can_issue_more = 1` got lost even in our old branch. I added it back in.

MichielDerhaeg replied:
I removed this in the original set of upstream patches because when you have 2 consequetive XBFU fusions:

```
slli <- cycle 1
srli <- cycle 1
slli <- cycle 1
srli <- cycle 2
```

The scheduler will start a new cycle after the second `slli` even though the hardware can run all 4 instructions in a single cycle.
This is fixed when these lines are removed, but it causes regressions elsewhere.
This is not a problem downstream because the xbfu fusion is still represented as a single instruction.

 I suggest we keep it like this (current state) upstream and fix it later.

luismgsilva committed:
RISC-V: Implement riscv_macro_fusion_pair_p for ARC-V RHX-100 series

Add the RHX-100 macro-op fusion cases to the table in riscv-fusion.cc
and enable them from the tune's fusible_ops field.  Also split BFEXT into
SRLI / SRAI.

Pairs that compete with adjacent load/store (ls-update, lui-st,
li-store) bail out when the next insn is a better adjacent memop
candidate.  Those, plus adjacent load/store themselves, wait until after
reload.

Fused_MAC only matches when the product is rs1 of the add:
  mul rd,...; add rd,rd,rs

I1->I2 means fusion only in that order; I1<->I2 means the two
instructions are interchangeable with regards to order.

RHX-100 enables:

  RISCV_FUSE_MULT_ADD       (riscv_fuse_mult_add, I1->I2)
      I1: mul rd,...; I2: add rd,rd,rs3 reading the mul result in rd
      (on RV64, mulw+addw with the result sign-extended to DImode).
  RISCV_FUSE_LI_BRANCH      (riscv_fuse_li_branch, I1->I2)
      I1: li rd; I2: conditional branch that uses rd.
  RISCV_FUSE_LI_STORE       (riscv_fuse_li_store, I1->I2)
      I1: li rt; I2: store of rt.
  RISCV_FUSE_LUI_ST         (riscv_fuse_lui_st, I1<->I2)
      I1/I2: lui and store.
  RISCV_FUSE_LUI_LD_REV     (riscv_fuse_lui_ld_reversible, I1<->I2)
      I1/I2: independent load and lui with distinct destinations.
  RISCV_FUSE_LS_UPDATE      (riscv_fuse_ls_update, I1<->I2)
      I1/I2: load or store with an addi that updates the access base.
  RISCV_FUSE_BFEXT_SRLI     (riscv_fuse_bfext_srli, I1->I2)
      I1: slli rd,rs; I2: srli rd,rd (bitfield extract).
  RISCV_FUSE_ADJACENT_LOAD  (riscv_fuse_adjacent_load, I1<->I2)
      I1/I2: two loads, same base, offsets |off1-off2| == XLEN/8.
  RISCV_FUSE_ADJACENT_STORE (riscv_fuse_adjacent_store, I1<->I2)
      I1/I2: two stores, same base, offsets |off1-off2| == XLEN/8.

Also rename RISCV_FUSE_BFEXT to RISCV_FUSE_BFEXT_SRLI and add
RISCV_FUSE_BFEXT_SRAI for other tunes.

gcc/ChangeLog:

	* config/riscv/riscv-fusion.cc (LUI_NONZERO_OPERAND): New macro.
	(riscv_fusion_enabled_p): Make non-static.
	(riscv_adjacent_memops_p): New function.
	(riscv_defer_for_adjacent_memop_p): New function.
	(riscv_ls_update_pair_p): New function.
	(riscv_fuse_lui_ld_pair_p): New function.
	(riscv_fuse_lui_ld): Use riscv_fuse_lui_ld_pair_p.
	(riscv_fuse_bfext_shift_p): New function.
	(riscv_fuse_bfext_srli): New function.
	(riscv_fuse_bfext_srai): New function.
	(riscv_fuse_mult_add): New function.
	(riscv_fuse_li_branch): New function.
	(riscv_fuse_adjacent_load): New function.
	(riscv_fuse_adjacent_store): New function.
	(riscv_fuse_ls_update): New function.
	(riscv_lui_st_pair_p): New function.
	(riscv_fuse_lui_st): New function.
	(riscv_fuse_li_store): New function.
	(riscv_fuse_lui_ld_reversible): New function.
	* config/riscv/riscv-protos.h (enum riscv_fusion_pairs): Rename
	RISCV_FUSE_BFEXT to RISCV_FUSE_BFEXT_SRLI; add the RHX-100 fusion
	bits and RISCV_FUSE_BFEXT_SRAI.
	(riscv_fusion_enabled_p): Declare.
	* config/riscv/riscv.cc (arcv_rhx100_tune_info): Set fusible_ops.
	* config/riscv/riscv.h (TARGET_ARCV_RHX100): New.

gcc/testsuite/ChangeLog:

	* gcc.target/riscv/fusion-adjacent-load.c: New test.
	* gcc.target/riscv/fusion-adjacent-store.c: New test.
	* gcc.target/riscv/fusion-bfext-srai.c: New test.
	* gcc.target/riscv/fusion-bfext.c: New test.
	* gcc.target/riscv/fusion-li-branch.c: New test.
	* gcc.target/riscv/fusion-li-store.c: New test.
	* gcc.target/riscv/fusion-ls-update-2.c: New test.
	* gcc.target/riscv/fusion-ls-update-3.c: New test.
	* gcc.target/riscv/fusion-ls-update-4.c: New test.
	* gcc.target/riscv/fusion-ls-update-5.c: New test.
	* gcc.target/riscv/fusion-ls-update.c: New test.
	* gcc.target/riscv/fusion-lui-ld.c: New test.
	* gcc.target/riscv/fusion-lui-st.c: New test.
	* gcc.target/riscv/fusion-mult-add.c: New test.

Co-authored-by: Artemiy Volkov <artemiyv@acm.org>
Co-authored-by: Michiel Derhaeg <michiel.derhaeg@globalfoundries.com>
Co-authored-by: Loeka Rogge <loeka.rogge@globalfoundries.com>
Signed-off-by: Luis Silva <luis.silva@globalfoundries.com>

luismgsilva committed:
RISC-V: Implement TARGET_SCHED_FUSION_PRIORITY for ARC-V RHX-100 series

To take better advantage of double load/store fusion, make use of the
sched_fusion pass that assigns unique "fusion priorities" to load/store
instructions and schedules operations on adjacent addresses together.
This maximizes the probability that loads/stores are fused with each
other rather than with other instructions.

The ARC-V specific helpers live in a new arcv.cc; the generic wrapper
riscv_sched_fusion_priority in riscv-fusion.cc dispatches to it and
falls back to a default priority for other tunes.

gcc/ChangeLog:

	* config.gcc: Build arcv.o.
	* config/riscv/riscv-fusion.cc (riscv_sched_fusion_priority): New.
	(riscv_adjacent_memops_p): Call arcv_pair_fusion_mode_allowed_p.
	* config/riscv/riscv-protos.h (riscv_sched_fusion_priority): Declare.
	(arcv_pair_fusion_mode_allowed_p): Declare.
	(arcv_sched_fusion_priority): Declare.
	* config/riscv/riscv.cc (TARGET_SCHED_FUSION_PRIORITY): Define.
	* config/riscv/t-riscv: Build arcv.o.
	* config/riscv/arcv.cc: New file.

Co-authored-by: Artemiy Volkov <artemiyv@acm.org>
Co-authored-by: Michiel Derhaeg <michiel.derhaeg@globalfoundries.com>
Signed-off-by: Luis Silva <luis.silva@globalfoundries.com>

luismgsilva committed:
RISC-V: Implement scheduling for ARC-V RHX-100 series

This patch implements instruction scheduling support for the dual-issue
ARC-V RHX-100 processor by adding scheduler hooks and state tracking for
the two execution pipes.

These hooks address microarchitectural details not covered by the
pipeline description.

The riscv_sched_variable_issue () and riscv_sched_reorder2 () hooks work
together to make sure that:
  - the critical path and the instruction priorities are respected;
  - both pipes are filled (taking advantage of parallel dispatch within
    the microarchitectural constraints);
  - there is as much fusion going on as possible;
  - the existing fusion pairs are not broken up.

riscv_sched_adjust_priority () slightly bumps the priority of load/store
pairs.  As a result it becomes easier for riscv_sched_reorder2 () to
schedule instructions in the memory pipe.

TARGET_SCHED_REORDER2 is defined for all RISC-V tunes, so
riscv_sched_reorder2 must preserve the remaining issue count for non-RHX
(via cached_can_issue_more from riscv_sched_variable_issue).

gcc/ChangeLog:

	* config/riscv/arcv.cc (struct arcv_sched_state): New.
	(arcv_sched_init): New function.
	(arcv_sched_reorder2): New function.
	(arcv_sched_adjust_priority): New function.
	(arcv_sched_adjust_cost): New function.
	(arcv_can_issue_more_p): New function.
	(arcv_sched_variable_issue): New function.
	* config/riscv/riscv-protos.h (arcv_sched_init): Declare.
	(arcv_sched_reorder2): Declare.
	(arcv_sched_adjust_priority): Declare.
	(arcv_sched_adjust_cost): Declare.
	(arcv_can_issue_more_p): Declare.
	(arcv_sched_variable_issue): Declare.
	* config/riscv/riscv.cc (cached_can_issue_more): New.
	(riscv_sched_init): Call arcv_sched_init.
	(riscv_sched_variable_issue): Handle RHX-100; cache the returned
	issue count.
	(riscv_sched_adjust_cost): Handle RHX-100.
	(riscv_sched_adjust_priority): New function.
	(riscv_sched_reorder2): New function; return cached_can_issue_more
	for non-RHX.
	(TARGET_SCHED_ADJUST_PRIORITY): Define.
	(TARGET_SCHED_REORDER2): Define.

gcc/testsuite/ChangeLog:

	* gcc.target/riscv/fusion-ls-update-2.c: Remove XFAIL, and fix the
	scan pattern.
	* gcc.target/riscv/fusion-ls-update.c: Remove XFAIL.

Co-authored-by: Artemiy Volkov <artemiyv@acm.org>
Co-authored-by: Michiel Derhaeg <michiel.derhaeg@globalfoundries.com>
Co-authored-by: Alex Turjan <alex.turjan@globalfoundries.com>
Co-authored-by: Loeka Rogge <loeka.rogge@globalfoundries.com>
Signed-off-by: Luis Silva <luis.silva@globalfoundries.com>

luismgsilva committed:
RISC-V: define_insn_and_split for multiply-add and bit-extract fusion

Introduce RTL patterns for multiply-add fusion on RV32.

Also cheapen zero_extract / sign_extract in riscv_rtx_costs when
RISCV_FUSE_BFEXT_SRLI / RISCV_FUSE_BFEXT_SRAI is enabled.

gcc/ChangeLog:

	* config/riscv/riscv.cc (riscv_rtx_costs): Cheapen zero_extract /
	sign_extract when BFEXT_SRLI / BFEXT_SRAI is enabled.
	* config/riscv/riscv.md (maddhisi4): Emit separate mul/add when
	RISCV_FUSE_MULT_ADD, not TARGET_XTHEADMAC, and not TARGET_64BIT.
	(umaddhisi4): New expand.
	(madd_fused): New insn_and_split.

gcc/testsuite/ChangeLog:

	* gcc.target/riscv/fusion-bfext-2.c: New test.
	* gcc.target/riscv/fusion-limm-condbr.c: New test.
	* gcc.target/riscv/fusion-madd.c: New test.

Co-authored-by: Artemiy Volkov <artemiyv@acm.org>
Co-authored-by: Michiel Derhaeg <michiel.derhaeg@globalfoundries.com>
Signed-off-by: Luis Silva <luis.silva@globalfoundries.com>

MichielDerhaeg commented on gcc/config/riscv/riscv-fusion.cc:
```diff
@@ -691,6 +954,270 @@ riscv_fuse_b_alui (rtx_insn *prev, rtx_insn *curr)
+    return true;
+
+  return false;
+}
```
Sorry, I was the person that asked for this to be split. But our lui_ld fusion requires the destination registers to be different, but the one that is already upstream requires them to be the same, so we should merge them again I think and give it some other name.

luismgsilva replied:
I do get a decrease in performance from `6.972939` to `6.941197` merging as it was.

EDIT: See https://github.com/foss-for-synopsys-dwc-arc-processors/gcc/commit/3131bd33143885449ae4f84402e8bb91fd4cf61a

MichielDerhaeg commented on gcc/config/riscv/riscv-fusion.cc:
```diff
@@ -741,6 +1268,22 @@ static const struct riscv_fusion_entry riscv_fusion_table[] =
     riscv_fuse_bfext, "RISCV_FUSE_BFEXT" },
```
Also, BFEXT is not entirely correct, it matches too much. Jin Ma's patch checks something similar.

luismgsilva committed:
tmp: Commit messages

Signed-off-by: Luis Silva <luis.silva@globalfoundries.com>

MichielDerhaeg commented on commit-msg/0001-RISC-V-Implement-riscv_macro_fusion_pair_p-for-ARC-V:
```diff
@@ -0,0 +1,96 @@
+Co-authored-by: Artemiy Volkov <artemiyv@acm.org>
+Co-authored-by: Michiel Derhaeg <michiel.derhaeg@globalfoundries.com>
+Signed-off-by: Luis Silva <luis.silva@globalfoundries.com>
+
```
Loeka needs to creditted

MichielDerhaeg commented on commit-msg/0001-RISC-V-Implement-riscv_macro_fusion_pair_p-for-ARC-V:
```diff
@@ -0,0 +1,96 @@
+	* gcc.target/riscv/arcv-fusion-mult-add-op0-dump.c: New test.
+	* gcc.target/riscv/arcv-fusion-mult-add-op1-dump.c: New test.
+	* gcc.target/riscv/arcv-fusion-shift-bitextract-dump.c: New test.
+	* gcc.target/riscv/arcv-fusion-store-lui-dump.c: New test.
```
I put it on the confluence page, and it is a bit of nit. But I would simplify the test names.
just call it `arcv-fusion-<fusion_name>-<number>.c` I don't know where the `dump` comes from, that's not really gcc convention I think.

Ah, also, not call it `arcv-`? what are the other fusion tests called?

luismgsilva replied:
See https://github.com/foss-for-synopsys-dwc-arc-processors/gcc/pull/296

MichielDerhaeg commented on commit-msg/0001-RISC-V-Implement-riscv_macro_fusion_pair_p-for-ARC-V:
```diff
@@ -0,0 +1,96 @@
+This patch adds the macro-op fusion cases used by the ARC-V RHX-100 and
+enables them in the tune_info (fusible_ops).  Fusion rules are collected in a
+table and dispatched from riscv_macro_fusion_pair_p; each is gated by its own
+RISCV_FUSE_* flag so a tune enables only the pairs its hardware supports.
```
> Fusion rules are collected in a
> table and dispatched from riscv_macro_fusion_pair_p; each is gated by its own
> RISCV_FUSE_* flag so a tune enables only the pairs its hardware supports.

Is obvious IMO, can be removed.

MichielDerhaeg commented on commit-msg/0001-RISC-V-Implement-riscv_macro_fusion_pair_p-for-ARC-V:
```diff
@@ -0,0 +1,96 @@
+Adjacent load/store fusion is prioritized: the other pairs are suppressed
+until sched2 so double load/store opportunities are not broken up.
+
+Fusion cases enabled for the RHX-100 (I1 -> I2):
```
You still reference I1/2 from the spec, but not in any of the descriptions.

MichielDerhaeg commented on commit-msg/0001-RISC-V-Implement-riscv_macro_fusion_pair_p-for-ARC-V:
```diff
@@ -0,0 +1,96 @@
+  RISCV_FUSE_ADJACENT_LOAD  (riscv_fuse_adjacent_load)
+      two loads, same base, offsets |off1-off2| == XLEN/8.
+  RISCV_FUSE_ADJACENT_STORE (riscv_fuse_adjacent_store)
+      two stores, same base, offsets |off1-off2| == XLEN/8.
```
We need to indicate which ones are reversible.

MichielDerhaeg commented on commit-msg/0001-RISC-V-Implement-riscv_macro_fusion_pair_p-for-ARC-V:
```diff
@@ -0,0 +1,96 @@
+riscv_fuse_bfext and riscv_fuse_lui_ld already existed; this patch adds the
+other eight checkers and enables the full set above for the RHX-100.  The
+memory pairs only fire after reload (reload_completed && !sched_fusion), which
+is why they are deferred to sched2.
```
Unfortunately not true

MichielDerhaeg commented on commit-msg/0001-RISC-V-Implement-riscv_macro_fusion_pair_p-for-ARC-V:
```diff
@@ -0,0 +1,96 @@
+  - Restructured the v1 memop+arith / memop+lui helpers into the more
+    specific LS_UPDATE, LUI_ST and LUI_LD_REV checkers.
+  - Dropped the config.gcc / t-riscv / arcv.o plumbing (no new file).
+  - Added 12 dump-scan tests (none in v1).
```
Also mention that many checks were added (like `REG_P`) that were forgotten before.

MichielDerhaeg commented on commit-msg/0003-RISC-V-Implement-scheduling-for-ARC-V-RHX-100-series:
```diff
@@ -0,0 +1,85 @@
+	that RHX-100 scheduling is implemented.
+	* gcc.target/riscv/arcv-fusion-ls-update-rev-dump.c: Likewise, and
+	fix the scan-rtl-dump pattern.
+
+Co-authored-by: Artemiy Volkov <artemiyv@acm.org>
+Co-authored-by: Michiel Derhaeg <michiel.derhaeg@globalfoundries.com>
+Co-authored-by: Alex Turjan <alex.turjan@globalfoundries.com>
+Signed-off-by: Luis Silva <luis.silva@globalfoundries.com>
```
Loeka needs credits here also.

MichielDerhaeg commented on commit-msg/0003-RISC-V-Implement-scheduling-for-ARC-V-RHX-100-series:
```diff
@@ -0,0 +1,85 @@
+    umbrella and the riscv_fusion_enabled_p forward declaration in
+    riscv.cc.
+  - Removed the custom arcv_next_fusible_insn helper and use the generic
+    next_nonnote_nondebug_insn_bb instead.
```
to exactly match the scheduler.

MichielDerhaeg commented on commit-msg/0003-RISC-V-Implement-scheduling-for-ARC-V-RHX-100-series:
```diff
@@ -0,0 +1,85 @@
+    fusion rework.
+  - arcv_sched_adjust_priority and arcv_sched_variable_issue now use
+    next_nonnote_nondebug_insn_bb + single_set instead of next_insn +
+    INSN_P, dropping the manual DEBUG/USE/CLOBBER guards.
```
you can merge that with the other comment about this, just mention we switched to this function and it is more correct. as it matches what the scheduler does.

MichielDerhaeg commented on commit-msg/0003-RISC-V-Implement-scheduling-for-ARC-V-RHX-100-series:
```diff
@@ -0,0 +1,85 @@
+    defined by the fusion patch, so this patch no longer touches riscv.h.
+  - Un-XFAIL'd arcv-fusion-ls-update-dump.c and
+    arcv-fusion-ls-update-rev-dump.c now that RHX-100 scheduling is
+    implemented (no test changes in v1).
```
Similar to the other patch, not enough checking was done previously and it should now be more correct.

MichielDerhaeg commented on commit-msg/0004-RISC-V-Add-instruction-patterns-for-32-bit-multiply-:
```diff
@@ -0,0 +1,54 @@
+From: Luis Silva <luis.silva@globalfoundries.com>
+Date: Mon, 13 Jul 2026 13:46:34 +0100
+Subject: [PATCH 4/4] RISC-V: Add instruction patterns for 32-bit multiply-add
+ and bit-extract fusion.
```
even the title needs to changed here

MichielDerhaeg commented on gcc/config/riscv/arcv.cc:
```diff
@@ -404,13 +404,6 @@ arcv_sched_variable_issue (rtx_insn *insn, int more)
-      sched_state.alu_pipe_scheduled_p = 1;
-      more -= 1;
-    }
-
```
Oops, added to the wrong commit.

MichielDerhaeg commented on gcc/config/riscv/iterators.md:
```diff
@@ -225,6 +225,8 @@
 (define_code_attr extract_shift [(sign_extract "ashiftrt")
 				 (zero_extract "lshiftrt")])
+(define_code_attr is_zero_extract [(sign_extract "false")
+				   (zero_extract "true")])
```
Forgot to remove this.

MichielDerhaeg committed:
Enable RTL checking for RISC-V

MichielDerhaeg commented on gcc/config/riscv/riscv.md:
```diff
@@ -4831,14 +4831,22 @@
   "&& 1"
   [(const_int 0)]
   "{
+     rtx addend = operands[3];
```
This change regresses performance somewhat but it is required for correctness.

The regression is not caused by this pattern, but other inaccuracies in the scheduling.
I suggest to postpone dealing with that regression for now.

luismgsilva commented on gcc/config/riscv/riscv.cc:
```diff
@@ -4710,7 +4714,9 @@ riscv_rtx_costs (rtx x, machine_mode mode, int outer_code, int opno ATTRIBUTE_UN
     case SIGN_EXTRACT:
-      if (TARGET_XTHEADBB && outer_code == SET
+      if ((TARGET_XTHEADBB
+	   || riscv_fusion_enabled_p (RISCV_FUSE_BFEXT))
```
This change triggered these failures:
```
FAIL: gcc.target/riscv/ext-dce-promote-2.c   -O0   scan-assembler \\tsrliw\t
FAIL: gcc.target/riscv/ext-dce-promote-2.c   -O1   scan-assembler \\tsrliw\t
FAIL: gcc.target/riscv/ext-dce-promote-2.c   -O2   scan-assembler \\tsrliw\t
FAIL: gcc.target/riscv/ext-dce-promote-2.c   -O3 -g   scan-assembler \\tsrliw\t
FAIL: gcc.target/riscv/ext-dce-promote-2.c   -Os   scan-assembler \\tsrliw\t
FAIL: gcc.target/riscv/ext-dce-promote-2.c  -Og -g   scan-assembler \\tsrliw\t
FAIL: gcc.target/riscv/ext-dce-promote-2.c  -Oz   scan-assembler \\tsrliw\t
```

This change lowered the RTL cost of `zero_extract` \ `sign_extract` whenever `RISCV_FUSE_BFXT` is enabled. With RHX as default mtune, combine prefers extract-shaped RTL over the srliw paths in riscv.md. The test case requires `srliw`.

luismgsilva commented on gcc/config/riscv/riscv.cc:
```diff
@@ -16537,6 +16582,9 @@ riscv_memtag_tag_bitsize ()
 #define TARGET_SCHED_MACRO_FUSION_PAIR_P riscv_macro_fusion_pair_p
 
+#undef TARGET_SCHED_FUSION_PRIORITY
+#define TARGET_SCHED_FUSION_PRIORITY riscv_sched_fusion_priority
```
We turned on GCC's sched_fusion pass for RISC-V by implementing `TARGET_SCHED_FUSION_PRIORITY`. The `interrupt-misaligned.c` test case checks that interrupt prologue/epilogue use a fixed order of stack saves (sd/sw/fsd at specific offsets). The test already disables normal instruction scheduling (-fno-schedule-insns, -fno-schedule-insns2) but not sched fusion, so the new pass reordered those memory operations.

I would add ` -fno-schedule-fusion` to the test case also.

MichielDerhaeg commented on gcc/config/riscv/riscv.cc:
```diff
@@ -980,7 +980,11 @@ static const struct riscv_tune_param arcv_rhx100_tune_info = {
+  (RISCV_FUSE_MULT_ADD | RISCV_FUSE_LI_BRANCH
+   | RISCV_FUSE_ADJACENT_LOAD | RISCV_FUSE_ADJACENT_STORE
+   | RISCV_FUSE_LS_UPDATE | RISCV_FUSE_LUI_ST
+   | RISCV_FUSE_LI_STORE | RISCV_FUSE_BFEXT
```
Regarding BFEXT. I suggest we simply split it into 2 separate fusions: one with srli and one with srai.

luismgsilva replied:
Initial PR https://github.com/foss-for-synopsys-dwc-arc-processors/gcc/pull/305

luismgsilva commented on gcc/config/riscv/riscv-protos.h:
```diff
@@ -867,11 +867,21 @@ enum riscv_fusion_pairs
+  RISCV_FUSE_ADJACENT_STORE = (1 << 17),
+  RISCV_FUSE_LS_UPDATE = (1 << 18),
+  RISCV_FUSE_LUI_ST = (1 << 19),
+  RISCV_FUSE_LI_STORE = (1 << 20),
```
@MichielDerhaeg Should we rename `RISCV_FUSE_LI_STORE` to `RISCV_FUSE_LI_ST` to align with others ?

exur00 committed:
riscv github runner use sourceware git://

The previously used https:// rate-limits github runners.
The git:// endpoint doesn't seem to do so.

luismgsilva commented:
V2 Patches have been sent upstream: https://gcc.gnu.org/pipermail/gcc-patches/2026-August/726434.html
